src/NServiceBus.Core/Pipeline/Incoming/IncomingContext.cs

```diff
@@ -18,22 +17,26 @@ public abstract class IncomingContext : BehaviorContext, IMessageProcessingConte
         /// <summary>
         /// Initializes a new instance of <see cref="IncomingContext" />.
         /// </summary>
-        protected IncomingContext(BehaviorContext parentContext)
+        protected IncomingContext(string messageId, string replyToAddress, IReadOnlyDictionary<string, string> headers, BehaviorContext parentContext)
             : base(parentContext)
         {
+
+            this.MessageId = messageId;
+            this.ReplyToAddress = replyToAddress;
+            this.MessageHeaders = headers;
         }
 
         /// <inheritdoc />
         public ContextBag Extensions => this;
 
         /// <inheritdoc />
-        public string MessageId => Get<IncomingMessage>().MessageId;
+        public string MessageId { get; }
 
         /// <inheritdoc />
-        public string ReplyToAddress => Get<IncomingMessage>().GetReplyToAddress();
+        public string ReplyToAddress { get; }
```

Comment:
Since ReplyToAddress is optional (send only endpoints won't have it and we also contemplate not providing it on events either) should we change this to a `TryGet`?